README.md
# YATA - Yay! Another Text Archive
Lightweight tool to archive text files - designed to help move files between VM/370 mini disks and Linux/Windows.

# Usage
    Usage   :   yata [options]
    Options :
      -h        Prints help message
      -v        Prints Version
      -x        Extract from Archive
      -c        Create Archive
      -d TEXT   Drive or Directory of files (default: A in CMS, . in Windows/Linux)
      -f TEXT   Archive file (default: archive.yata)

# YATA Archive file format
The format is trivial. The YATA archive file is text based, variable length lines up to 80 characters (so it can go through punch card files). 

The key assumption is that ASCII/EBCDIC/Unicode conversions are handled by the transport layer not YATA.

The first character is a line control character
- File Start (+)

    +filename.filetype   

- Line Start (\>) first 79 characters

    \>linetext

- Line Continue (\<) next 79 characters etc.

    \<linetext

- Archive End (*)

    \*

Trailing spaces are removed. On CMS flies will need to have recfm / lrecl reset (which I don't see as an issue)

# VMARC Archiving
This is the standard tool for VM/370 archiving, and it has GUI tools, etc. 
I suggest you use this for archives that you want to archive!
I am not attempting to replace VMARC but I wanted a really lightweight tool to transfer source directories, hence YATA.


history.txt
YATA - Yay! Another Text Archive
================================
Adrian Sutherland - adrian@sutherlandonline.org

History

1.2.5 - 5 April 2021
F0011 - VM/370 CE Support - Specifically CMSINSTALL is deprecated and the
                            binary file just has the "statically" linked
                            module. Simplify ...
        CMS - Disable Random R/W Buffer
        Default Archive renamed to archive.yata
        .gitignore updated for CLION

1.2.4 - 18 Oct 2020
F0010 - Update Build Scripts
      - Added MacOS Target
      - Added PEG file type
      - Removed some minor warnings
      - Trim (and replace trimmed) white space from archive lines

1.2.3 - 25 Jul 2020
F0009 - Performance Improvements
        Build on "builder" tag
1.2.2 - 23 Jul 2020
        CMS Build fix (wrong version of GCCLIB used in 1.2.1)
1.2.1 - 22 Jul 2020
FOOO8 - Link against GCCLIB 0.7.12 / VM370 Image v1.4.17
      - CMS workarounds removed (GCCLIB Updates)
      - Added TMPFTYPE whitelist file type

1.2.0 - 21 Feb 2020
FOOO7 - Link against GCCLIB 0.7.1 / VM370 Image v1.4.0

1.1.3 - 20 Feb 2020
FOOO6 - Added HELPCMD HELPCMD2

1.1.2 - 14 Feb 2020
F0005 - Suppress trailing blank line on CMS yet make
        sure all lines end in a NL (Linux/Windows)

1.1.1 - 12 Feb 2020
F0004 - CMS Install Script

1.1.0 - 11 Feb 2020
F0003 - Keep archive line length <=80 characters
F0002 - Automate CMS Build
F0001 - Autobuild (Windows and Ubuntu)

< 1.0.2 - Called Spltcon - Prehistory versions!

spltcon.c
/***********************************************************************/
/*                                                                     */
/* SPLTCON.C                                                           */
/*                                                                     */
/* CONCAT [d] - Combines file on d drive (default a) into SPLIT CONCAT */
/* SPLIT [d] - Splits files from SPLIT CONCAT A in to seperata files   */
/*             on d drive (default A)                                  */
/*                                                                     */
/***********************************************************************/
#ifdef _WIN32
#define _CRT_SECURE_NO_WARNINGS
#endif
 
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
 
#define MAXRECL 800
 
#ifdef __CMS
 
#define HELPTEXT "\nSPLIT|CONCAT [D], where D (default=A) is the disk\n"
#define SPLIT "SPLIT"
#define CONCAT "CONCAT"
 
#include <cmssys.h>
#define CONCATFILE "SPLIT CONCAT A1"
char* includeTypes[] = { "C", "H", "EXEC", "ASSEMBLE", "LISTING", "COPY", "MACLIB",
                         "MACRO", "PARM", "MEMO"};
int toupper(int c);
int tolower(int c);
 
#else
 
#define HELPTEXT "\nsplit|concat [D], where D (default is . [i.e. current]) is the directory\n"
 
#ifdef _WIN32
 
#include <windows.h>
 
#else
 
#include <dirent.h>
#include <libgen.h>
 
#endif
 
#define SPLIT "split"
#define CONCAT "concat"
 
#include <ctype.h>
#define CONCATFILE "spltcon.txt"
char* includeTypes[] = { "c", "h", "exec", "assemble", "listing", "copy", "maclib",
                         "macro", "parm", "memo"};
 
#endif
 
int concat(void);
int split(void);
char* validateFileName(char *listFileLine);
char* toStoredName(char *fileName);
char* fromStoredName(char *fileName);
char *trimTrailingSpace(char *str);
 
char* drive;
 
#ifdef _WIN32
char* basename(char* path)
{
    char* base = strrchr(path, '\\');
    return base ? base + 1 : path;
}
#endif
 
int main(int argc, char * argv[]) {
  char* name;
  int error = 0;
  printf("SPLTCON VERSION 1.0.2\n");
 
#ifdef __CMS
  name = argv[0];
#else
  name = basename(argv[0]);
  name = strtok(name, ".");
#endif
 
  if (argc>2) {
    error = 1;
    printf("ERROR: Too many arguments\n");
  }
  if (argc==2) {
#ifdef __CMS
    if (strlen(argv[1])!=1) {
      error = 1;
      printf("ERROR: Invalid drive\n");
    }
#endif
    drive = argv[1];
  }
  else {
#ifdef __CMS
    drive = "A";
#else
    drive = ".";
#endif
  }
 
  if (!strcmp(name,SPLIT)) {
    if (!error) return split();
  }
  else if (!strcmp(name,CONCAT)) {
    if (!error) return concat();
  }
  else printf("ERROR: Module name must be SPLIT or CONCAT - not %s\n",name);
  printf(HELPTEXT);
  return -1;
}
 
int concat(void) {
#ifdef __CMS
  int stackSize;
  int files;
  char consoleLine[131];
#else
#ifdef _WIN32
  WIN32_FIND_DATA fdFile;
  HANDLE hFind = NULL;
#else
  DIR *d;
  struct dirent *dir;
#endif
#endif
  char* fileName;
  FILE* inFile;
  FILE* outFile;
  char lineBuffer[MAXRECL];
 
  outFile = fopen(CONCATFILE, "w");
 
#ifdef __CMS
  int i;
  char command[40];
  stackSize = CMSstackQuery();
  sprintf(command, "LISTFILE * * %s (FORMAT STACK", drive);
  CMScommand(command, CMS_COMMAND);
  files = CMSstackQuery() - stackSize;
  for (i=0; i<files; i++) {
    if (CMSconsoleRead(consoleLine)) {
      fileName = validateFileName(consoleLine);
#else
 
#ifdef _WIN32
 
 char dirAndName[260];
 snprintf(dirAndName, 259, "%s\\*.*", drive);
 hFind = FindFirstFile(dirAndName, &fdFile);
  if (hFind != INVALID_HANDLE_VALUE) {
    do {
      snprintf(dirAndName, 259, "%s\\%s", drive, fdFile.cFileName);
      fileName = validateFileName(dirAndName);
 
#else
  d = opendir(drive);
  if (d)  {
    char dirAndName[260];
    while ((dir = readdir(d)) != NULL) {
      snprintf(dirAndName, 259, "%s/%s", drive, dir->d_name);
      fileName = validateFileName(dirAndName);
 
#endif
#endif
 
      if (fileName) {
        inFile = fopen(fileName,"r");
        if (inFile==NULL) {
          printf("ERROR: Can not open file %s\n", fileName);
          return -1;
        }
        fprintf(outFile, "+%s\n", toStoredName(fileName));
 
        while (fgets(lineBuffer, MAXRECL , inFile) != NULL) {
          trimTrailingSpace(lineBuffer);
          fprintf(outFile, ">%s\n", lineBuffer);
        }
        fclose(inFile);
      }
    }
#ifdef _WIN32
    while (FindNextFile(hFind, &fdFile) != 0);
    FindClose(hFind);
#endif
  }
 
#ifdef __CMS
#else
#ifdef _WIN32
#else
  closedir(d);
#endif
#endif
 
  fclose(outFile);
 
  return 0;
}
 
char* validateFileName(char *listFileLine) {
  int i;
  /* TODO Check if the file is not a directory */
 
#ifdef __CMS
 
  int recl;
  char trimmedFileType[9];
 
  listFileLine[20] = 0;   /* 20 is the end of the file mode */
 
  /* If the file record length is too long skip */
  listFileLine[29] = 0;
  recl  = atoi(listFileLine+24);
  if (recl>MAXRECL) {
   return NULL;
  }
 
  /* position 9 is where the file type starts */
  for (i=9; listFileLine[i] && listFileLine[i]!=' '; i++) {
    trimmedFileType[i-9] = listFileLine[i];
  }
  trimmedFileType[i-9] = 0;
 
#else
 
  char buffer[100];
  strncpy(buffer, listFileLine, 100);
  buffer[99]=0;
  (void)strtok(buffer, ".");
  char *trimmedFileType = strtok(NULL, " .");
  if (trimmedFileType == NULL) {
    printf("WARNING: file %s skipped\n", listFileLine);
    return NULL;
  }
 
#endif
 
  /* Only process if the file is one of the included types */
  int noTypes = sizeof(includeTypes) / sizeof(includeTypes[0]);
 
  for (i=0; i<noTypes; ++i)
  {
    if (!strcmp(includeTypes[i], trimmedFileType)) {
      return listFileLine;
    }
  }
  printf("WARNING: file %s skipped\n", listFileLine);
  return NULL;
}
 
int split(void) {
  char* fileName;
  FILE* inFile;
  FILE* outFile = NULL;
  char lineBuffer[MAXRECL];
 
  inFile = fopen(CONCATFILE, "r");
  if (inFile==NULL) {
    printf("ERROR: Error opening file %s\n", CONCATFILE);
    return -1;
  }
  while (fgets(lineBuffer, MAXRECL , inFile) != NULL) {
    switch (lineBuffer[0]) {
      case '+':
        if (outFile) {
          fclose(outFile);
        }
        fileName = fromStoredName(lineBuffer+1);
        outFile = fopen(fileName, "w");
        if (outFile==NULL) {
          fclose(inFile);
          printf("ERROR: Error opening file %s\n", fileName);
          return -1;
        }
        break;
 
      case '>':
        if (!outFile) {
          printf("ERROR: Output file not specified on CONCAT file");
          fclose(inFile);
          return -1;
        }
        trimTrailingSpace(lineBuffer);
#ifdef __CMS
        /* This is a work around for a bug in CMSSYS where it does
           not like writing a line with just a \n */
        if (strlen(lineBuffer) > 1) {
          fprintf(outFile, "%s\n", lineBuffer+1);
        }
        else {
          fputs(" \n", outFile); /* Have to add a space :-( */
        }
#else
        fprintf(outFile, "%s\n", lineBuffer+1);
#endif
        break;
 
      default:
        if (outFile) {
          fclose(outFile);
        }
        fclose(inFile);
        printf("ERROR: Invalid character in col 0\n");
        return -1;
     }
  }
  if (outFile) {
    fclose(outFile);
  }
  fclose(inFile);
  return 0;
}
 
char* toStoredName(char *fileName) {
#ifdef __CMS
 static char name[25];
 int i;
 char *fn;
 char *ft;
 fn = strtok(fileName, " ");
 ft = strtok(NULL, " ");
 
 sprintf(name, "%s.%s",fn,ft);
 for (i=0; name[i]; i++) name[i] = tolower(name[i]);
 return name;
#else
 return basename(fileName);
#endif
}
 
char* fromStoredName(char *fileName) {
#ifdef __CMS
 static char name[25];
 int i;
 char *fn = fileName;
 char *ft;
 
 for (ft=fn; *ft; ft++) if (*ft=='.') {
   *ft=0; ft++;
   break;
 }
 TrimTrailingSpace(ft);
 
 if (strlen(fn) > 8) fn[8]=0;
 if (strlen(ft) > 8) ft[8]=0;
 
 sprintf(name, "%s %s %s", fn, ft, drive);
 for (i=0; name[i]; i++) name[i] = toupper(name[i]);
 return name;
#else
 static char name[100];
 fileName[ strlen(fileName)-1 ]=0; /* Remove new-line */
 snprintf(name, 99, "%s/%s", drive, fileName);
 return name;
#endif
}
 
 
char *trimTrailingSpace(char *str)
{
  char *end;
 
  end = str + strlen(str) - 1;
  while ( end >= str &&
          ( *end == ' ' || *end == '\n' || *end == '\t')
        ) end--;
 
  /* Terminate */
  end[1] = 0;
 
  return str;
}
 
#ifdef __CMS
 
static const unsigned char lower[] = "abcdefghijklmnopqrstuvwxyz";
static const unsigned char upper[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
 
int toupper(int c) {
  unsigned char *p;
  p=strchr(lower,c);
  if (p) return upper[p-lower];
  return c;
}
 
int tolower(int c) {
  unsigned char *p;
  p=strchr(upper,c);
  if (p) return lower[p-upper];
  return c;
}
 
#endif


cmsbuild.sh
#!/bin/sh
# Make YATA on CMS

# Exit if there is an error
set -e

# IPL
herccontrol -v
herccontrol "ipl 141" -w "USER DSC LOGOFF AS AUTOLOG1"
herccontrol "/cp start c" -w "RDR"

# LOGON CMSUSER
herccontrol "/cp disc" -w "^VM/370 Online"
herccontrol "/logon cmsuser cmsuser" -w "^CMS"
herccontrol "/" -w "^Ready;"

# Read MAKE EXEC
herccontrol -m >tmp; read mark <tmp; rm tmp
echo "USERID  CMSUSER\n:READ  MAKE     EXEC    " > tmp
cat make.exec >> tmp
netcat -q 0 localhost 3505 < tmp
rm tmp
herccontrol -w "HHCRD012I" -f $mark 
herccontrol "/" -w "RDR FILE"
herccontrol "/read *" -w "^Ready;"

# Read YATA C
herccontrol -m >tmp; read mark <tmp; rm tmp
echo "USERID  CMSUSER\n:READ  YATA     C       " > tmp
cat yata.c >> tmp
netcat -q 0 localhost 3505 < tmp
rm tmp
herccontrol -w "HHCRD012I" -f $mark 
herccontrol "/" -w "RDR FILE"
herccontrol "/read *" -w "^Ready;"

# Read ARCHIVE YATA (for testing)
herccontrol -m >tmp; read mark <tmp; rm tmp
echo "USERID  CMSUSER\n:READ  ARCHIVE  YATA    " > tmp
cat test/archive.yata >> tmp
netcat -q 0 localhost 3505 < tmp
rm tmp
herccontrol -w "HHCRD012I" -f $mark 
herccontrol "/" -w "RDR FILE"
herccontrol "/read *" -w "^Ready;"

# Build
herccontrol "/make" -w "^Ready;"

# Sanity test
herccontrol "/yata -x" -w "^Ready;"
herccontrol "/listf test* exec a" -w "^Ready;"

# Make and load Tape
herccontrol "/cp disc" -w "^VM/370 Online"
herccontrol "/logon operator operator" -w "RECONNECTED AT"
hetinit -n -d yatabin.aws
herccontrol "devinit 480 io/yatabin.aws" -w "^HHCPN098I"
herccontrol "/attach 480 to cmsuser as 181" -w "TAPE 480 ATTACH TO CMSUSER"
herccontrol "/cp disc" -w "^VM/370 Online"

# Write to tape
herccontrol "/logon cmsuser cmsuser" -w "RECONNECTED AT"
herccontrol "/begin"
herccontrol "/tape dump yata module a" -w "^Ready;"
herccontrol "/detach 181" -w "^Ready;"

# LOGOFF
herccontrol "/logoff" -w "^VM/370 Online"

# SHUTDOWN
herccontrol "/logon operator operator" -w "RECONNECTED AT"
herccontrol "/shutdown" -w "^HHCCP011I"


cmsinstall.sh
#!/bin/sh
# **************************************************************************
# * THIS FILE IS DEPRECATED and is not designed to work on the VM/370 CE ***
# * Going forward installation is a responsibility of the distribution not *
# * of the package provider                                                *
# **************************************************************************
#
# Install YATA on CMS
#


# Exit if there is an error
set -e

# IPL
herccontrol -v
herccontrol "ipl 141" -w "USER DSC LOGOFF AS AUTOLOG1"

# LOGON MAINT AND READ TAPE
herccontrol "/cp disc" -w "^VM/370 Online"
herccontrol "/logon maint cpcms" -w "^CMS VERSION"
herccontrol "/" -w "^Ready;"
herccontrol "devinit 480 io/yatabin.aws" -w "^HHCPN098I"
herccontrol "/attach 480 to maint as 181" -w "TAPE 480 ATTACH TO MAINT"
herccontrol "/access 19e c" -w "^Ready;"
herccontrol "/tape load * * c" -w "^Ready;"
herccontrol "/rename yata module c1 = = c2"  -w "^Ready;"
herccontrol "/detach 181" -w "^Ready;"
herccontrol "/release c"  -w "^Ready;"
herccontrol "/ipl 190" -w "^CMS VERSION"
herccontrol "/savesys cms" -w "^CMS VERSION"
herccontrol "/" -w "^Ready;"
herccontrol "/logoff" -w "^VM/370 Online"

# LOGON CMSUSER
herccontrol "/logon cmsuser cmsuser" -w "^CMS VERSION"
herccontrol "/" -w "^Ready;"

# Sanity test
herccontrol "/yata -v" -w "^Ready;"

# LOGOFF
herccontrol "/logoff" -w "^VM/370 Online"

# SHUTDOWN
herccontrol "/logon operator operator" -w "RECONNECTED AT"
herccontrol "/shutdown" -w "^HHCCP011I"


yata.c
/***********************************************************************/
/*                                                                     */
/* YATA.C                                                              */
/* Yay! Another Text Archive. Lightweight tool to archive text files   */
/*                                                                     */
/* Adrian Sutherland                                                   */
/*                                                                     */
/***********************************************************************/

#ifdef _WIN32
#define _CRT_SECURE_NO_WARNINGS
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXRECL 800
#define ARCLINELEN 80
#define VERSION "1.2.5"

#ifdef __CMS

#include <cmssys.h>
static char* includeTypes[] = { "C", "H", "EXEC", "ASSEMBLE", "LISTING", 
              "COPY", "MACLIB", "TMPFTYPE", "PEG", "RXAS", "Y", "RE", "TXT",
             "MACRO", "PARM", "MEMO", "HELPCMD", "HELPCMD2" };
#define ARCHIVE "ARCHIVE YATA A1"
#define DRIVE "A"
#define FILENAMELEN 25
static char fileNameBuffer[FILENAMELEN];

#else

#ifdef _WIN32

#include <windows.h>

#else

#include <dirent.h>
#include <libgen.h>

#endif

#include <ctype.h>
static char* includeTypes[] = { "c", "h", "exec", "assemble", "listing",
              "copy", "maclib", "tmpftype","peg", "rxas", "y", "re", "txt",
             "macro", "parm", "memo", "helpcmd", "helpcmd2" };
#define ARCHIVE "archive.yata"
#define DRIVE "."

#endif

static char* toUpperString(char* string);
static int create_archive();
static int extract_archive();
static char* validateFileName(char* listFileLine);
static char* toStoredName(char* fileName);
static char* fromStoredName(char* fileName);
static char* trimTrailingSpace(char* str);
static char* trimTrailingSpaceAndFixLength(char* str, unsigned int len);
static char* trimTrailingSpaceWithLength(char* str, unsigned int len);


#ifdef _WIN32
static char* basename(char* path)
{
  char* base = strrchr(path, '\\');
  return base ? base + 1 : path;
}
#endif

static char* drive;
static char* archive_file;

static void help() {
  char* helpMessage =
    "YATA (Yay! Another Text Archive) - "
    "Lightweight tool to archive text files\n"
    "Version :   " VERSION "\n"
    "Usage   :   yata [options]\n"
    "Options :\n"
    "  -h        Prints help message\n"
    "  -v        Prints Version\n"
    "  -x        Extract from Archive\n"
    "  -c        Create Archive\n"
    "  -d TEXT   Drive or Directory of files (default: A in CMS,"
    " . in Windows/Linux)\n"
    "  -f TEXT   Archive file (default: archive.yata)\n";

#ifdef __CMS
  helpMessage = toUpperString(helpMessage);
#endif        
  printf("%s",helpMessage);
}

static void error_and_exit(int rc, char* message) {

#ifdef __CMS
  message = toUpperString(message);
#endif        

  fprintf(stderr, "ERROR: %s\n", message);
  help();
  exit(rc);
}

int main(int argc, char* argv[]) {
  int extract_mode = -1;
  int i;
  drive = DRIVE;
  archive_file = ARCHIVE;

  /* Parse arguments - DIY style! */
  for (i = 1; i < argc && argv[i][0] == '-'; i++) {
    if (strlen(argv[i]) > 2) {
      error_and_exit(2, "Invalid argument");
    }
    switch (toupper((argv[i][1]))) {
    case '-':
      break;

    case 'X': /* Extract*/
      extract_mode = 1;
      break;

    case 'C': /* Create */
      extract_mode = 0;
      break;

    case 'D': /* Directory / Drive */
      i++;
      if (i >= argc) {
        error_and_exit(2, "Missing drive/directory after -d");
      }
      drive = argv[i];
      break;

    case 'F': /* Archive file */
#ifdef __CMS
      fileNameBuffer[0] = 0;
      for (i++; i < argc && argv[i][0] != '-'; i++) {
        strncat(fileNameBuffer, " ", FILENAMELEN - strlen(fileNameBuffer)
          - 1);
        strncat(fileNameBuffer, argv[i], FILENAMELEN - strlen(fileNameBuffer)
          - 1);
      }
      i--;
      if (strlen(fileNameBuffer) == 0) {
        error_and_exit(2, "Missing filename after -f");
      }
      archive_file = fileNameBuffer + 1;
#else
      i++;
      if (i >= argc) {
        error_and_exit(2, "Missing filename after -f");
      }
      archive_file = argv[i];
#endif
      break;

    case 'V': /* Version */
      printf("%s\n", VERSION);
      exit(0);

    case 'H': /* Help */
      help();
      exit(0);

    default:
      error_and_exit(2, "Invalid argument");
    }
  }

  if (i < argc) {
    error_and_exit(2, "Unexpected Arguments");
  }

  if (extract_mode == -1) {
    error_and_exit(2, "Create or eXtract must be specified");
  }

  if (extract_mode) {
    return extract_archive();
  }
  return create_archive();
}

static int create_archive() {
#ifdef __CMS
  int stackSize;
  int files;
  char consoleLine[131];
#else
#ifdef _WIN32
  WIN32_FIND_DATA fdFile;
  HANDLE hFind = NULL;
#else
  DIR* d;
  struct dirent* dir;
#endif
#endif
  char* fileName;
  FILE* inFile;
  FILE* outFile;
  char lineBuffer[MAXRECL];

  outFile = fopen(archive_file, "w");

#ifdef __CMS
  int i;
  char command[40];
  setbuf(outFile, 0); /* Disables te overhead of the random r/w cache */
  stackSize = CMSstackQuery();
  sprintf(command, "LISTFILE * * %s (FORMAT STACK", drive);
  CMScommand(command, CMS_COMMAND);
  files = CMSstackQuery() - stackSize;
  for (i = 0; i < files; i++) {
    if (CMSconsoleRead(consoleLine)) {
      fileName = validateFileName(consoleLine);
#else

#ifdef _WIN32

  char dirAndName[260];
  snprintf(dirAndName, 259, "%s\\*.*", drive);
  hFind = FindFirstFile(dirAndName, &fdFile);
  if (hFind != INVALID_HANDLE_VALUE) {
    do {
      snprintf(dirAndName, 259, "%s\\%s", drive, fdFile.cFileName);
      fileName = validateFileName(dirAndName);

#else
  d = opendir(drive);
  if (d) {
    char dirAndName[260];
    while ((dir = readdir(d)) != NULL) {
      snprintf(dirAndName, 259, "%s/%s", drive, dir->d_name);
      fileName = validateFileName(dirAndName);

#endif
#endif

      if (fileName) {
        inFile = fopen(fileName, "r");
        if (inFile == NULL) {
          printf("ERROR: Can not open file %s\n", fileName);
          return 1;
        }
#ifdef __CMS
        setbuf(inFile, 0); /* Disables te overhead of the random r/w cache */
#endif
        fprintf(outFile, "+%s\n", toStoredName(fileName));
        while (fgets(lineBuffer, MAXRECL, inFile) != NULL) {
          trimTrailingSpace(lineBuffer);
          char* line = lineBuffer;
          if (strlen(line) > ARCLINELEN - 1) {
            trimTrailingSpaceWithLength(line,79);
            fprintf(outFile, ">%.*s\n", ARCLINELEN - 1, line);
            line += 79;
            do {
              if (strlen(line) > 79) {
                trimTrailingSpaceWithLength(line,79);
                fprintf(outFile, "<%.*s\n", ARCLINELEN - 1, line);
                line += 79;
              }
              else {
                fputc('<', outFile);
                fputs(line, outFile);
                fputc('\n', outFile);
                break;
              }
            } while (1);
          }
          else {
            fputc('>', outFile);
            fputs(line, outFile);
            fputc('\n', outFile);
          }
        }
        fclose(inFile);
      }
    }
#ifdef _WIN32
    while (FindNextFile(hFind, &fdFile) != 0);
    FindClose(hFind);
#endif
  }

#ifdef __CMS
#else
#ifdef _WIN32
#else
  closedir(d);
#endif
#endif
  fprintf(outFile, "*\n");
  fclose(outFile);

  return 0;
}

static char* validateFileName(char* listFileLine) {
  int i;
  /* TODO Check if the file is not a directory */

#ifdef __CMS

  int recl;
  char trimmedFileType[9];

  listFileLine[20] = 0;   /* 20 is the end of the file mode */

  /* If the file record length is too long skip */
  listFileLine[29] = 0;
  recl = atoi(listFileLine + 24);
  if (recl > MAXRECL) {
    return NULL;
  }

  /* position 9 is where the file type starts */
  for (i = 9; listFileLine[i] && listFileLine[i] != ' '; i++) {
    trimmedFileType[i - 9] = listFileLine[i];
  }
  trimmedFileType[i - 9] = 0;

#else

  char buffer[100];
  strncpy(buffer, listFileLine, 100);
  buffer[99] = 0;
  (void)strtok(buffer, ".");
  char* trimmedFileType = strtok(NULL, " .");
  if (trimmedFileType == NULL) {
    printf("WARNING: file %s skipped\n", listFileLine);
    return NULL;
  }

#endif

  /* Only process if the file is one of the included types */
  int noTypes = sizeof(includeTypes) / sizeof(includeTypes[0]);

  for (i = 0; i < noTypes; ++i)
  {
    if (!strcmp(includeTypes[i], trimmedFileType)) {
      return listFileLine;
    }
  }
  printf("WARNING: file %s skipped\n", listFileLine);
  return NULL;
}

static void writeLine(FILE* outFile, char* lineBuffer) {
  trimTrailingSpace(lineBuffer);
  fputs(lineBuffer, outFile);
  fputc('\n', outFile);
}

static int extract_archive() {
  int lineNo = 0;
  char* fileName;
  FILE* inFile;
  FILE* outFile = NULL;
  char lineBuffer[MAXRECL + 1];
  char line[ARCLINELEN + 3]; /* could add /r /n /0 */
  int needToWriteLine = 0;
  int startedArchive = 0;

  inFile = fopen(archive_file, "r");
  if (inFile == NULL) {
    printf("ERROR: Error opening archive file %s\n", archive_file);
    return 1;
  }
#ifdef __CMS
    setbuf(inFile, 0); /* Disables te overhead of the random r/w cache */
#endif
    while (fgets(line, ARCLINELEN + 3, inFile) != NULL) {
    lineNo++;
    trimTrailingSpace(line);
    if (strlen(line) > 80) {
      if (startedArchive) {
        if (outFile) {
          if (needToWriteLine)
            writeLine(outFile, lineBuffer);
          fclose(outFile);
        }
        fclose(inFile);
        printf("ERROR: Line %d, Line >80 chars \"%s\"\n", lineNo, line);
        return 1;
      }
    }

    switch (line[0]) {
    case '*':
      startedArchive = 1;
      break;

    case '+':
      startedArchive = 1;
      if (outFile) {
        if (needToWriteLine) {
          if (strlen(lineBuffer) > 0) {
            writeLine(outFile, lineBuffer);
          }
          needToWriteLine = 0;
          lineBuffer[0] = 0;
        }
        fclose(outFile);
      }
      fileName = fromStoredName(line + 1);
      outFile = fopen(fileName, "w");
      if (outFile == NULL) {
        fclose(inFile);
        printf("ERROR: Error opening file %s\n", fileName);
        return 1;
      }
#ifdef __CMS
      setbuf(outFile, 0); /* Disables te overhead of the random r/w cache */
#endif
      break;

    case '>':
        trimTrailingSpaceAndFixLength(line, 80);
      startedArchive = 1;
      if (!outFile) {
        printf("ERROR: Output file not specified, error in archive file");
        fclose(inFile);
        return 1;
      }
      if (needToWriteLine) {
        writeLine(outFile, lineBuffer);
      }
      strncpy(lineBuffer, line + 1, MAXRECL - 1);
      needToWriteLine = 1;
      break;

    case '<':
        trimTrailingSpaceAndFixLength(line, 80);
      startedArchive = 1;
      if (!needToWriteLine) {
        printf("ERROR: Append line without a line, error in archive file");
        fclose(inFile);
        return 1;
      }
      strncat(lineBuffer, line + 1, MAXRECL - strlen(lineBuffer) - 1);
      break;

    default:
      /* Skip any junk before the start */
      if (startedArchive) {
        if (outFile) {
          if (needToWriteLine)
            writeLine(outFile, lineBuffer);
          fclose(outFile);
        }
        fclose(inFile);
        printf("ERROR: Line %d, Invalid character in col 0 of \"%s\"\n",
          lineNo, line);
        return 1;
      }
    }
  }

  if (outFile) {
    if (needToWriteLine) {
      if (strlen(lineBuffer) > 0) {
        writeLine(outFile, lineBuffer);
      }
    }
    fclose(outFile);
  }
  fclose(inFile);
  return 0;
}

static char* toStoredName(char* fileName) {
#ifdef __CMS
  static char name[25];
  int i;
  char* fn;
  char* ft;
  fn = strtok(fileName, " ");
  ft = strtok(NULL, " ");

  sprintf(name, "%s.%s", fn, ft);
  for (i = 0; name[i]; i++) name[i] = tolower(name[i]);
  return name;
#else
  return basename(fileName);
#endif
}

static char* fromStoredName(char* fileName) {
#ifdef __CMS
  static char name[25];
  int i;
  char* fn = fileName;
  char* ft;

  for (ft = fn; *ft; ft++) if (*ft == '.') {
    *ft = 0; ft++;
    break;
  }

  if (strlen(fn) > 8) fn[8] = 0;
  if (strlen(ft) > 8) ft[8] = 0;

  sprintf(name, "%s %s %s", fn, ft, drive);
  for (i = 0; name[i]; i++) name[i] = toupper(name[i]);
  return name;
#else
  static char name[100];
  snprintf(name, 99, "%s/%s", drive, fileName);
  return name;
#endif
}

static char* trimTrailingSpace(char* str)
{
    return trimTrailingSpaceWithLength(str,strlen(str));
}

/* Nulls trailing spaces */
static char* trimTrailingSpaceWithLength(char* str, unsigned int len)
{
    char* end;

    end = str + len - 1;
    while ( end >= str &&
          (*end == 0 || *end == ' ' || *end == '\n' ||
           *end == '\t' || *end == '\r') ) {
        *end = 0;
        end--;
    }

    return str;
}

static char* trimTrailingSpaceAndFixLength(char* str, unsigned int len)
{
    unsigned int l, i;

    trimTrailingSpaceWithLength(str,strlen(str));

    /* Add spaces so the line is len chars long */
    l = strlen(str);
    for (i=l; i<len; i++) str[i] = ' ';
    str[i] = 0;

    return str;
}

static char* toUpperString(char* string) {
  int i;
  for (i = 0; string[i]; i++) string[i] = (char)toupper(string[i]);
  return string;
}
